test/e2e/network/service.go

```diff
@@ -101,42 +101,87 @@ var (
+// The returned function returns polling interval and timeout and function returning
+// a set of hosts that were pinged by the checker.
+type affinityCheckerPollingArgs func() (time.Duration, time.Duration, func() []string)
+
+func affinityCheckFromPod(execPod *v1.Pod, serviceIP string, servicePort int) affinityCheckerPollingArgs {
+	timeout := AffinityTimeout
+	// interval considering a maximum of 2 seconds per connection
+	interval := 2 * AffinityConfirmCount * time.Second
+
+	serviceIPPort := net.JoinHostPort(serviceIP, strconv.Itoa(servicePort))
+	curl := fmt.Sprintf(`curl -q -s --connect-timeout 2 http://%s/`, serviceIPPort)
+	cmd := fmt.Sprintf("for i in $(seq 0 %d); do echo; %s ; done", AffinityConfirmCount, curl)
+	getHosts := func() []string {
+		stdout, err := framework.RunHostCmd(execPod.Namespace, execPod.Name, cmd)
+		if err != nil {
+			framework.Logf("Failed to get response from %s. Retry until timeout", serviceIPPort)
+			return nil
+		}
+		return strings.Split(stdout, "\n")
+	}
+
+	return func() (time.Duration, time.Duration, func() []string) {
+		return interval, timeout, getHosts
+	}
+}
+
+func affinityCheckFromTest(cs clientset.Interface, serviceIP string, servicePort int) affinityCheckerPollingArgs {
+	interval := 2 * time.Second
+	timeout := e2eservice.GetServiceLoadBalancerPropagationTimeout(cs)
+
+	params := &e2enetwork.HTTPPokeParams{Timeout: 2 * time.Second}
+	getHosts := func() []string {
+		var hosts []string
+		for i := 0; i < AffinityConfirmCount; i++ {
+			result := e2enetwork.PokeHTTP(serviceIP, servicePort, "", params)
+			if result.Status == e2enetwork.HTTPSuccess {
+				hosts = append(hosts, string(result.Body))
+			}
+		}
+		return hosts
+	}
+
+	return func() (time.Duration, time.Duration, func() []string) {
+		return interval, timeout, getHosts
+	}
+}
+
 // CheckAffinity function tests whether the service affinity works as expected.
 // If affinity is expected, the test will return true once affinityConfirmCount
 // number of same response observed in a row. If affinity is not expected, the
 // test will keep observe until different responses observed. The function will
 // return false only in case of unexpected errors.
 func checkAffinity(cs clientset.Interface, execPod *v1.Pod, serviceIP string, servicePort int, shouldHold bool) bool {
-	serviceIPPort := net.JoinHostPort(serviceIP, strconv.Itoa(servicePort))
-	curl := fmt.Sprintf(`curl -q -s --connect-timeout 2 http://%s/`, serviceIPPort)
-	cmd := fmt.Sprintf("for i in $(seq 0 %d); do echo; %s ; done", AffinityConfirmCount, curl)
-	timeout := AffinityTimeout
-	if execPod == nil {
-		timeout = e2eservice.GetServiceLoadBalancerPropagationTimeout(cs)
+	var pollingArgs affinityCheckerPollingArgs
+	if execPod != nil {
+		pollingArgs = affinityCheckFromPod(execPod, serviceIP, servicePort)
+	} else {
+		pollingArgs = affinityCheckFromTest(cs, serviceIP, servicePort)
 	}
+
+	interval, timeout, getHosts := pollingArgs()
```

Comment:
@danwinship - thanks a lot for the first pass of review; I refactored this test a bit and hopefully it's much cleaner now.
PTAL